subsys/net/ip/ipv4_autoconf.c

```diff
@@ -245,23 +253,44 @@ static void send_announcement(struct net_if *iface)
 	net_if_get_config(iface)->ipv4auto.state = NET_IPV4_AUTOCONF_ANNOUNCE;
 
 	k_delayed_work_submit(&net_if_get_config(iface)->ipv4auto.timeout,
-			      ipv4_autoconf_get_timeout(iface));
+			    ipv4_autoconf_get_timeout(iface));
 }
 
-void net_ipv4_autoconf_start(struct net_if *iface)
+static void if_handler(struct net_mgmt_event_callback *cb,
+		    u32_t mgmt_event,
+		    struct net_if *iface)
 {
+
+	if ((mgmt_event & NET_EVENT_IF_DOWN) &&
+		(net_if_get_config(iface)->ipv4auto.state == NET_IPV4_AUTOCONF_ASSIGNED)) {
+		net_if_ipv4_addr_rm(iface, &net_if_get_config(iface)->ipv4auto.current_ip);
+		return;
+	}
+
 	/* initialize interface and start probing */
```

Comment:
I think you need to also check NET_EVENT_IF_UP here as the event number is not a bitmask.